### PR TITLE
Issue 356: Rules that fail when using BigQuery UNNEST() with alias

### PR DESCRIPTION
**Pull request by barrywhart**
Addresses #356. This PR adds a new ANSI-level set called `value_table_functions`. It's empty at the ANSI level, but BigQuery updates it to add `UNNEST`. This set is not used in the grammars, but when there is a `table_expression` with an alias, affected rules use this set to determine the nature of that alias: Is it a table alias or a column alias? This is a weird concept, but that's essentially what BigQuery does.

Context: `UNNEST()` is usually, but not always, used to unpack an array of scalar values. In this case, it returns a "table" with only one column. BigQuery has the concept of ["value tables"](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#value_tables):
>In BigQuery, a value table is a table where the row type is a single value. In a regular table, each row is made up of columns, each of which has a name and a type. In a value table, the row type is just a single value, and there are no column names.

In this situation, BugQuery seems to sort of treat the "table" and its single column as the same thing, allowing the alias for the value table to be used directly as a column. It's fairly natural for users of `UNNEST()`, but it's pretty weird from a SQLFluff perspective, since the nature of the alias has to be deduced with more complex logic. And, as you'll see in some of my comments on the issue (#356), it's probably not perfect for two reasons, both covered in the BigQuery docs [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#value_tables):
* While `UNNEST()` is usually used with scalar arrays, BigQuery also supports arrays of `STRUCT`, which return regular tables, not value tables, when unnested.
* BigQuery supports **implicit** unnesting of array columns, in which case, the `UNNEST()` function does not appear in the query.

This PR makes no attempt to address either of these last two challenges, but it does address all the issues initially raised in #356, which I think will cover the majority of array unnesting use cases.

### CODE REVIEW

File: src/sqlfluff/core/dialects/dialect_ansi.py [hunk 1]
```diff
@@ -142,6 +142,17 @@
...
 )
 
+# Set the value table functions. These are functions that, if they appear as
+# an item in "FROM', are treated as returning a COLUMN, not a TABLE. Among
+# Apparently, among dialects supported by SQLFluff, only BigQuery has this
+# concept, but this set is defined in the ANSI dialect because:
+# - It impacts core linter rules (see L020 and several other rules that subclass
+#   from it) and how they interpret the contents of table_expressions
+# - At least one other database (DB2) has the same value table function,
+#   UNNEST(), as BigQuery. DB2 is not currently supported by BigQuery.
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/dialects/dialect_ansi.py:152]: `not currently supported by sqlfluff`?

File: src/sqlfluff/core/dialects/dialect_ansi.py [hunk 2]
```diff
@@ -142,6 +142,17 @@
     ]
 )
 
+# Set the value table functions. These are functions that, if they appear as
+# an item in "FROM', are treated as returning a COLUMN, not a TABLE. Among
```
Comment (alanmcruickshank) [inline src/sqlfluff/core/dialects/dialect_ansi.py:146]: Trailing `Among`.

**Review by alanmcruickshank — APPROVED** [review]
Couple of nits on your docs :)

The rest looks good though. I'm not sure I 100% get the concept, but your explanation was clear enough that I think I know what's going on.

**Comment by barrywhart** [comment]
@alanmcruickshank and @NiallRees: This PR is ready for review. Alan's initial take on the issue was that it may be a parsing problem, but it's something more subtle (and a little odd) which required refactoring several affected rules. I think I made some reasonable choices in terms of complexity and handling something that is currently dialect specific but may not always be. 🤞 

I'm very interested in your feedback about the design decisions as well as the PR as a whole. Thanks!

File: src/sqlfluff/core/dialects/dialect_ansi.py (hunk 1 above)
Comment (barrywhart) [reply src/sqlfluff/core/dialects/dialect_ansi.py:152]: Ha, good catch. Fingers faster than 🧠. 😄

**Pull request merged**